jetty-client/src/main/java/org/eclipse/jetty/client/HttpReceiver.java

```diff
@@ -591,46 +647,151 @@ public String toString()
         FAILURE
     }
 
-    private class Decoder extends IteratingNestedCallback
+    private class ContentListeners
+    {
+        private final Map<Object, Long> demands = new ConcurrentHashMap<>();
+        private final LongConsumer demand = HttpReceiver.this::demand;
+        private final List<Response.DemandedContentListener> listeners;
+
+        private ContentListeners(List<Response.ResponseListener> responseListeners)
+        {
+            listeners = responseListeners.stream()
+                .filter(Response.DemandedContentListener.class::isInstance)
+                .map(Response.DemandedContentListener.class::cast)
+                .collect(Collectors.toList());
```

Comment:
Should there be a short circuit optimization for when there is only 1 listener (the most common case)? 
Edit: oh there is below.... but perhaps a final boolean would be better than checking list size on each call?

Reply:
Checking the list size of an array list is just a field read, like a reading the `boolean` field would be, but using less memory (no need of the `boolean` field).